File: python/31-40/Sudoku Solver.py
class Solution:
    def solve(self,board,row,col):
        for i in range(0,9):
            for j in range(0,9):
                cell = board[i][j]
                if cell == '.':
                    for can in range(1,10):
                        value = str(can)
                        if(self.valid(board,i,j,value)):
                            board[i][j] = value
                            if self.solve(board,i,j):
                                return True
                            board[i][j]='.'
                    return False
        return True
    
    def valid(self,board,i,j,value):
        #check row,col
        for t in range(0,9):
            if board[t][j]==value or board[i][t]==value:
                return False
        #check sqar
        row,col = i//3*3,j//3*3
        for tr in range(row,row+3):
            for tc in range(col,col+3):
                if board[tr][tc]==value:
                    return False
        return True
    # @param board, a 9x9 2D array
    # Solve the Sudoku by modifying the input board in-place.
    # Do not return any value.
    def solveSudoku(self, board):
        self.solve(board,0,0)

File: Cpp/31-40/Sudoku Solver.cpp
/*
    runtime: 40ms
    error: 0

*/
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        dfs(board,0,0);
    }
    bool dfs(vector<vector<char> >& board , int row, int col){
        for(int i=row; i< 9; i++){
            for(int j=0; j<9; j++){
                if(board[i][j]=='.'){
                    for(int k =1; k<=9; k++){
                        if(isvalid(board,i,j,k)){
                            board[i][j] = '0'+k;
                            if(dfs(board,i,j)) return true;
                        }
                    }
                    board[i][j]='.';
                    return false;
                }
            }
        }
        return true;
    }
    bool isvalid(vector<vector<char> > & board, int row, int col, int val){
          char ch = '0'+val;
          for(int j=0; j<9; j++){
              if(board[row][j]==ch)
                return false;
          }
        //check row
        for(int i=0; i<9; i++){
            if(board[i][col]==ch)
              return false;
          }
        //check square
            int prow = row/3;
            int pcol = col/3;
            for(int i = prow*3; i<prow*3+3; i++){
              for(int j = pcol*3; j <pcol*3+3; j++){
                if(board[i][j]==ch)
                  return false;
              }
            }
          return true;
    }
};
    
